src/core/textlayout.js

```diff
@@ -0,0 +1,176 @@
+           createPromiseCapability, QuadTree */
+ 
+'use strict';
+ 
+var TextLayoutEvaluator = (function TextLayoutEvaluatorClosure() {
+  function TextLayoutEvaluator() {
+  }
+ 
+  var NonWhitespaceRegexp = /\S/;
+  function isAllWhitespace(str) {
+    return !NonWhitespaceRegexp.test(str);
+  }
+
+  // Trying to minimize Date.now() usage and check every 100 time 
+  var TIME_SLOT_DURATION_MS = 20;
+  var CHECK_TIME_EVERY = 100;
+  function TimeSlotManager() {
+    this.reset();
+  }
+  TimeSlotManager.prototype = {
+    check: function TimeSlotManager_check() {
+      if (++this.checked < CHECK_TIME_EVERY) {
+        return false;
+      }
+      this.checked = 0;
+      return this.endTime <= Date.now();
+    },
+    reset: function TimeSlotManager_reset() {
+      this.endTime = Date.now() + TIME_SLOT_DURATION_MS;
+      this.checked = 0;
+    }
+  };
+  var deferred = Promise.resolve();
+
+  TextLayoutEvaluator.prototype = {
+    overlapBy : function (min1, min2, max1, max2, by) {
+        var d = Math.min(max1,max2) - Math.max(min1,min2);
+        return d > (max1-min1) * by || d > (max2-min2) * by;
+    },
+    
+    addToQuadTree: function(o, i, styles) {
+      var style = styles[o.fontName];
+      if (isAllWhitespace(o.str)) {
+        // Whitespace elements aren't visible, but they're used for copy/paste.
+        o.isWhitespace = true;
+      } else {
+        // Line numbers can mess up flow, so we detect and handle them them.
+        var n = Number(o.str.replace(/^\s+|\s+$/g,''));
+        if(n % 1 === 0) {
+          o.asInt = n;
+        }
+      }
+      var tx = o.transform;
+      
+      var angle = Math.atan2(tx[1], tx[0]);
+      if (style.vertical) {
+        angle += Math.PI / 2;
+      }
+      var fontHeight = Math.sqrt((tx[2] * tx[2]) + (tx[3] * tx[3]));
+      var fontAscent = fontHeight;
+      if (style.ascent) {
+        fontAscent = style.ascent * fontAscent;
+      } else if (style.descent) {
+        fontAscent = (1 + style.descent) * fontAscent;
+      }
+
+      // Set y,x to the bottom left. They're the smaller values.
+      o.y = tx[5];
+      o.x = (angle === 0) ? tx[4] : (tx[4] + (fontAscent * Math.sin(angle)));
+      o.vertical = style.vertical;
+      
+      o.id = i;         // Used to uniquely identify object.
+      o.right = null;   // The nearest object to the right.
+      o.bottom = null;  // The nearest object to the left.
+          
+      this.quadtree.insert(o);
+    },
+    
+    calculateTextFlow: function (bounds, objs, styles) {
+      var timeSlotManager = new TimeSlotManager();
+      var self = this;
+      self.bounds = bounds;
+      
+      // Put everything into the quadtree for O(logN) positional look ups.
+      self.quadtree = new QuadTree(bounds, 4, 16);
+      for (var i = 0, len = objs.length; i < len; i++) {
+        self.addToQuadTree(objs[i], i, styles);
+      }
+      
+      // The top item has a big impact on the flow, so track it.
+      var top_obj = null;
```

Comment:
Nit: `top_obj` -> `topObj`. Using underscore is reserved for constants, please see https://github.com/mozilla/pdf.js/wiki/Style-Guide#naming.